Mathlib/Analysis/SpecialFunctions/Elliptic/Weierstrass.lean

```diff
@@ -0,0 +1,277 @@
+
+## tags
+
+Weierstrass p-functions, Weierstrass p functions
+
+-/
+
+@[expose] public section
+
+open Module
+
+noncomputable section
+
+/-- A pair of `ℝ`-linearly independent complex numbers.
+They span the period lattice in `PeriodPair.lattice`,
+and are the periods of the elliptic functions we shall construct. -/
+structure PeriodPair : Type where
+  /-- The first period in a `PeriodPair`. -/
+  ω₁ : ℂ
+  /-- The first period in a `PeriodPair`. -/
+  ω₂ : ℂ
+  indep : LinearIndependent ℝ ![ω₁, ω₂]
+
+variable {M : Type*} [AddCommMonoid M] [TopologicalSpace M] (L : PeriodPair)
+
+/-- The `ℝ`-basis of `ℂ` determined by a pair of periods. -/
+def PeriodPair.basis : Basis (Fin 2) ℝ ℂ :=
+  basisOfLinearIndependentOfCardEqFinrank L.indep (by simp)
+
+@[simp] lemma PeriodPair.basis_zero : L.basis 0 = L.ω₁ := by simp [basis]
+@[simp] lemma PeriodPair.basis_one : L.basis 1 = L.ω₂ := by simp [basis]
+
+/-- The lattice spanned by a pair of periods. -/
+def PeriodPair.lattice : Submodule ℤ ℂ := Submodule.span ℤ {L.ω₁, L.ω₂}
+
+lemma PeriodPair.mem_lattice {L : PeriodPair} {x : ℂ} :
+    x ∈ L.lattice ↔ ∃ m n : ℤ, m * L.ω₁ + n * L.ω₂ = x := by
+  simp only [lattice, Submodule.mem_span_pair, zsmul_eq_mul]
+
+lemma PeriodPair.ω₁_mem_lattice : L.ω₁ ∈ L.lattice := Submodule.subset_span (by simp)
+lemma PeriodPair.ω₂_mem_lattice : L.ω₂ ∈ L.lattice := Submodule.subset_span (by simp)
+
+lemma PeriodPair.ω₁_div_two_notMem_lattice : L.ω₁ / 2 ∉ L.lattice := by
+  intro h
+  obtain ⟨m, n, e⟩ := mem_lattice.mp h
+  rw [← sub_eq_zero] at e
+  have := (LinearIndependent.pair_iff.mp L.indep (m - 1/2) n (by convert e using 1; simp; ring)).1
+  rw [sub_eq_zero, eq_div_iff (by norm_num)] at this
+  norm_cast at this
+  exact Int.two_dvd_ne_zero (n := 1).mpr rfl ⟨m, by linarith⟩
+
+lemma PeriodPair.ω₂_div_two_notMem_lattice : L.ω₂ / 2 ∉ L.lattice := by
+  intro h
+  obtain ⟨m, n, e⟩ := mem_lattice.mp h
+  rw [← sub_eq_zero] at e
+  have := (LinearIndependent.pair_iff.mp L.indep m (n - 1/2) (by convert e using 1; simp; ring)).2
+  rw [sub_eq_zero, eq_div_iff (by norm_num)] at this
+  norm_cast at this
+  exact Int.two_dvd_ne_zero (n := 1).mpr rfl ⟨n, by linarith⟩
+
+-- helper lemma to connect to the ZLattice API
+lemma PeriodPair.lattice_eq_span_range_basis :
+    L.lattice = Submodule.span ℤ (Set.range L.basis) := by
+  have : Finset.univ (α := Fin 2) = {0, 1} := rfl
+  rw [PeriodPair.lattice, ← Set.image_univ, ← Finset.coe_univ, this]
+  simp [Set.image_insert_eq]
+
+instance : DiscreteTopology L.lattice := L.lattice_eq_span_range_basis ▸ inferInstance
+
+instance : IsZLattice ℝ L.lattice := by
+  simp_rw [L.lattice_eq_span_range_basis]
+  infer_instance
+
+lemma PeriodPair.isClosed_lattice : IsClosed (X := ℂ) L.lattice :=
+  @AddSubgroup.isClosed_of_discrete _ _ _ _ _ L.lattice.toAddSubgroup
+    (inferInstanceAs (DiscreteTopology L.lattice))
+
+lemma PeriodPair.isClosed_lattice_sdiff (s : Set ℂ) : IsClosed (L.lattice \ s) := by
```

Comment:
Is it simpler to have this result, or the equivalent one that any subset of `L.lattice` is closed.